docs/rfc/rfc-025-support-app-side-mempool.md

```diff
@@ -0,0 +1,252 @@
+at a set of defined junctures and Tendermint would never add to it or remove
+from it without input from the application. Tendermint would impose _no_
+validity constraints on the contents of this list and would not attempt to
+remove items unless instructed to.
+
+### Mock API of the GossipList
+
+Outlined below is a proposed API for this data structure. These calls would be
+added to the ABCI API and would come to replace the current `CheckTx` call.
+
+#### `OfferPendingTransaction`
+
+`OfferPendingTransaction` replaces the `CheckTx` call that is invoked when
+Tendermint receives a submitted or gossiped transaction. The `GossipList` will
+invoke `OfferPendingTransaction` on _every_ transaction sent to Tendermint that
+does not match one of the transactions already in the `GossipList`. The mempool
+currently drops gossiped transactions before `CheckTx` is called if the
+transaction is considered invalid for a Tendermint-defined reason such as
+inclusion in the mempool 'cache' or it overflows the max transaction size.
+
+The application can indicate if the transaction should be added to the
+`GossipList` via `ResponseOfferPendingTransaction`'s `GossipStatus` field. If
+the `GossipList` is full, the application must list a transaction to remove from
+`GossipList`, otherwise the transaction will not be added. In this way,
+a transaction will _never_ leave the list unless the application removes it from
+the list explicitly.
+
+```proto
+message RequestOfferPendingTransaction {
+  bytes  tx = 1;
+  int64 gossip_list_max_size = 2;
+  int64 gossip_list_current_size = 3;
+}
+
+message ResponseOfferPendingTransaction {
+  GossipStatus  gossip_status = 1;
+  enum GossipStatus {
+    UNKNOWN  = 0;
+    GOSSIP = 1;
+    NO_GOSSIP = 2;
+  }
+  repeated bytes removals =2
+}
+```
+
+#### `UpdateTransactionGossipList`
+
+`UpdateTransactionGossipList` would be a simple method that allows the
+application to exactly set the contents of the `GossipList`. Tendermint would
+call `UpdateTransactionGossipList` on the application, which would respond with
+the list of all transactions to gossip. The contents of the `GossipList` would
+be completely replaced with the contents provided by the application in
+`UpdateTransactionGossipList`.
+
+```proto
+message UpdateTransactionGossipListRequest {
+  int64 max_size = 1; // application cannot provide more than `max_size` transactions.
+}
+
+message UpdateTransactionGossipListResponse {
+  repeated bytes = 1;
+}
+```
+
+This new `ABCI` method would serve multiple functions. First, it would replace
+the re-`CheckTx` calls made by Tendermint after each block is committed. After
+each block is committed, Tendermint currently passes the entire contents of the
+mempool to the application one-by-one via `CheckTx` calls with `CheckTxType` set
+to `RECHECK`. The application, in this way, can then inspect the entire mempool
+and remove any transactions that became invalid as a result of the most recent
+block being committed.
```

Comment:
Yes, that's correct. This would further enforce the notion that the application should be fully responsible for keeping track of what viable transactions exist at the moment because it will need to hand the full set to Tendermint after each block. This vision reduces Tendermint's relationship to the pending transactions to a very simple facility. All that Tendermint can and should do in relationship to pending transactions is to gossip them when told.